File: CPP/Problem_015.cpp
#include <iostream>
using namespace std;

unsigned long long nCr(int n, int r) {
    r = min(r, n - r);
    unsigned long long res = 1;
    for (int i = 1; i <= r; ++i) {
        res = res * (n - r + i) / i; // exact division each step
    }
    return res;
}

int main() {
    cout << nCr(40, 20) << "\n"; // 137846528820
    return 0;
}

File: Java/Problem_015.java
import java.math.BigInteger;

public class Main {
    private static BigInteger nCr(int n, int r) {
        r = Math.min(r, n - r);
        BigInteger res = BigInteger.ONE;

        for (int i = 1; i <= r; i++) {
            res = res.multiply(BigInteger.valueOf(n - r + i))
                     .divide(BigInteger.valueOf(i)); // exact division
        }
        return res;
    }

    public static void main(String[] args) {
        System.out.println(nCr(40, 20)); // 137846528820
    }
}
